### PR TITLE
Use pytest-xdist to run some pytest tests in parallel

### PR DESCRIPTION
**Pull request by definelicht**
- Use pytesdt-xdist to test in parallel.
- Introduce a new flag for generating the SDFG cache name with more options, allowing a random name based on the process ID to be used, avoiding clashes between parallel tests.
- Reorganize heterogeneous tests so they can potentially be run in parallel (if we add more runners).
- Rename a bunch of SDFGs to avoid name clashes even without using the unique cache name.

### CODE REVIEW

File: dace/config_schema.yml [hunk 1]
```diff
@@ -734,6 +734,23 @@ required:
...
+        type: str 
+        default: name 
+        title: Naming of cache entry 
+        description: >
+            Determine the name of the generated dacecache folder.
+            "name" uses the name of the SDFG, causing it to be overridden by
+            other programs using a similar SDFG name.
+            "hash" uses a mangled name using the hash of the SDFG, such that any
+            change to the SDFG will generate a different cache folder.
+            "unique" uses a name based on the currently running Pythion process
```
Comment (tbennun) [inline dace/config_schema.yml:747]:
```suggestion
            "unique" uses a name based on the currently running Python process 
```

File: dace/config_schema.yml [hunk 2]
```diff
@@ -734,6 +734,23 @@ required:
...
+        title: Naming of cache entry 
+        description: >
+            Determine the name of the generated dacecache folder.
+            "name" uses the name of the SDFG, causing it to be overridden by
+            other programs using a similar SDFG name.
+            "hash" uses a mangled name using the hash of the SDFG, such that any
+            change to the SDFG will generate a different cache folder.
+            "unique" uses a name based on the currently running Pythion process 
+            at code generation time, such that no caching or clashes can happen
+            between different processes or subsequent invokations of Python.
```
Comment (tbennun) [inline dace/config_schema.yml:749]:
```suggestion
            between different processes or subsequent invocations of Python.
```

File: tests/instrumentation_test.py [hunk 3]
```diff
@@ -32,6 +32,7 @@ def onetest(instrumentation: dace.InstrumentationType, size=128):
     C = np.zeros([size, size], dtype=np.float64)
 
     sdfg: dace.SDFG = slowmm.to_sdfg()
+    sdfg._name = f"instrumentation_test_{instrumentation.name}"
```
Comment (tbennun) [inline tests/instrumentation_test.py:35]:
```suggestion
    sdfg.name = f"instrumentation_test_{instrumentation.name}"
```

This should work too I think.

File: tests/transformations/loop_to_map_test.py [hunk 4]
```diff
@@ -66,8 +68,8 @@ def make_sdfg(with_wcr, map_in_guard, reverse_loop, use_variable, assign_after):
                                 """\
 static std::mutex mutex;
 std::unique_lock<std::mutex> lock(mutex);
-std::ofstream of("loop_to_map_test.txt", std::ofstream::app);
-of << i << "\\n";""",
+std::ofstream of("{}", std::ofstream::app);
```
Comment (tbennun) [inline tests/transformations/loop_to_map_test.py:71]: maybe fstring here?

File: tests/transformations/loop_to_map_test.py [hunk 5]
```diff
@@ -111,11 +119,6 @@ def apply_and_verify(sdfg, n):
 
     num_transformations = sdfg.apply_transformations(LoopToMap)
 
-    try:
-        os.remove("loop_to_map_test.txt")
```
Comment (tbennun) [inline tests/transformations/loop_to_map_test.py:115]: don't you want to still remove the temp file?

**Review by tbennun — CHANGES_REQUESTED** [review]

**Comment by tbennun** [comment]
Also, why is the coverage down?

**Comment by definelicht** [comment]
> 
> 
> Also, why is the coverage down?

The coverage is not down in this commit, but something might have happened in moving the FPGAs to the other testing system? Could you take a look at the github workflow file, and see if something is missing?

File: tests/transformations/loop_to_map_test.py (hunk 5 above)
Comment (definelicht) [reply tests/transformations/loop_to_map_test.py:115]: This is not necessary anymore, as we are using a unique `tempfile.NamedTemporaryFile` every time the function is called, which is cleaned up after the test.

File: dace/config_schema.yml (hunk 2 above)
Comment (definelicht) [reply dace/config_schema.yml:749]: why tf is it spelled "invoke", but "invocation"? English is stoopid

File: tests/transformations/loop_to_map_test.py (hunk 4 above)
Comment (definelicht) [reply tests/transformations/loop_to_map_test.py]: Done in 43cc83a7c74d7625c20fbe108e65e6ba32ea7ea9

File: tests/instrumentation_test.py (hunk 3 above)
Comment (definelicht) [reply tests/instrumentation_test.py]: Done in 2ec3478f917352bdf1816155caf1da3262b6b0c8

**Review by tbennun — APPROVED** [review]

**Comment by tbennun** [comment]
test-fpga doesn't have a "report coverage" clause

**Pull request merged**